wp-includes/mysql/class-mysql-lexer.php

```diff
@@ -0,0 +1,2752 @@
+			$this->text      = $matches[0];
+			$this->position += strlen( $this->text );
+			$this->c         = $this->input[ $this->position ] ?? null;
+			$this->n         = $this->input[ $this->position + 1 ] ?? null;
+
+			// When preceded by a dot, it is always an identifier.
+			if ( $p >= 0 && '.' === $this->input[ $p ] ) {
+				$this->type = self::IDENTIFIER;
+			} elseif ( '_' === $la && isset( self::UNDERSCORE_CHARSETS[ strtolower( $this->text ) ] ) ) {
+				$this->type = self::UNDERSCORE_CHARSET;
+			} else {
+				$this->identifier_or_keyword();
+			}
+		} elseif ( null === $la ) {
+			$this->match_eof();
+			$this->token_instance = new MySQL_Token( self::EOF, '<EOF>' );
+			return false;
+		} else {
+			$this->consume();
+			$this->type = self::INVALID_INPUT;
+		}
+
+		$this->token_instance = null === $this->type ? null : new MySQL_Token( $this->type, $this->text, $this->channel );
+		return true;
+	}
+
+	protected function la( int $i ) {
+		return $this->input[ $this->position + $i - 1 ] ?? null;
+	}
+
+	protected function consume() {
+		$this->text     .= $this->c;
+		$this->position += 1;
+		$this->c         = $this->input[ $this->position ] ?? null;
+		$this->n         = $this->input[ $this->position + 1 ] ?? null;
+	}
+
+	protected function match_eof() {
+		if ( null === $this->c ) {
+			$this->consume();
+		} else {
+			throw new RuntimeException( 'Current character is not EOF.' );
+		}
+	}
+
+	protected function identifier_or_keyword() {
+		$text = strtoupper( $this->get_text() );
+
+		// Lookup the string in the token table.
+		$this->type = self::TOKENS[ $text ] ?? self::IDENTIFIER;
+		if ( self::IDENTIFIER === $this->type ) {
+			return;
+		}
+
+		// Apply MySQL server version specifics (positive number: >= <version>, negative number: < <version>).
+		if ( isset( self::VERSIONS[ $this->type ] ) ) {
+			$version = self::VERSIONS[ $this->type ];
+			if ( $this->server_version < $version || -$version >= $this->server_version ) {
+				$this->type = self::IDENTIFIER;
+				return;
+			}
+		}
+
+		// Apply MySQL version ranges manually.
+		if (
+			self::MAX_STATEMENT_TIME_SYMBOL === $this->type
+			&& ( $this->server_version <= 50704 || $this->server_version >= 50708 )
+		) {
+			$this->type = self::IDENTIFIER;
+			return;
+		} elseif (
+			self::NONBLOCKING_SYMBOL === $this->type
+			&& ( $this->server_version <= 50700 || $this->server_version >= 50706 )
+		) {
+			$this->type = self::IDENTIFIER;
+			return;
+		} elseif (
+			self::REMOTE_SYMBOL === $this->type
+			&& ( $this->server_version < 80003 || $this->server_version >= 80014 )
+		) {
+			$this->type = self::IDENTIFIER;
+			return;
+		}
+
+		// Determine function calls.
+		if ( isset( self::FUNCTIONS[ $this->type ] ) ) {
+			// Skip any whitespace character if the SQL mode says they should be ignored.
+			$i = 1;
+			if ( $this->is_sql_mode_active( self::SQL_MODE_IGNORE_SPACE ) ) {
+				while ( $this->is_whitespace( $this->la( $i ) ) ) {
```

Comment:
nitpick: `strspn()` would be a lot faster than calling a few PHP functions for each byte